// @ngneat/hot-toast/esm2020/lib/hot-toast.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { DynamicViewModule } from '@ngneat/overview';
import { AnimatedIconComponent } from './components/animated-icon/animated-icon.component';
import { HotToastComponent } from './components/hot-toast/hot-toast.component';
import { CheckMarkComponent } from './components/indicator/icons/checkmark/checkmark.component';
import { ErrorComponent } from './components/indicator/icons/error/error.component';
import { LoaderComponent } from './components/indicator/icons/loader/loader.component';
import { IndicatorComponent } from './components/indicator/indicator.component';
import { ToastConfig } from './hot-toast.model';
import { HotToastContainerComponent } from './components/hot-toast-container/hot-toast-container.component';
import { WarningComponent } from './components/indicator/icons/warning/warning.component';
import { InfoComponent } from './components/indicator/icons/info/info.component';
import * as i0 from "@angular/core";
export class HotToastModule {
    static forRoot(config) {
        return {
            ngModule: HotToastModule,
            providers: [{ provide: ToastConfig, useValue: config }],
        };
    }
}
HotToastModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: HotToastModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HotToastModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.1", ngImport: i0, type: HotToastModule, declarations: [HotToastContainerComponent,
        HotToastComponent,
        AnimatedIconComponent,
        IndicatorComponent,
        CheckMarkComponent,
        ErrorComponent,
        LoaderComponent,
        WarningComponent,
        InfoComponent], imports: [CommonModule, DynamicViewModule] });
HotToastModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: HotToastModule, imports: [CommonModule, DynamicViewModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: HotToastModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        HotToastContainerComponent,
                        HotToastComponent,
                        AnimatedIconComponent,
                        IndicatorComponent,
                        CheckMarkComponent,
                        ErrorComponent,
                        LoaderComponent,
                        WarningComponent,
                        InfoComponent,
                    ],
                    imports: [CommonModule, DynamicViewModule],
                }]
        }] });
